### PR TITLE
[ISSUE #5754] [RIP-57] Add asynchronous interfaces to MessageStore

### PR DESCRIPTION
**Pull request by ShadowySpirits**
link to #5754

### CODE REVIEW

File: broker/src/main/java/org/apache/rocketmq/broker/processor/DefaultPullMessageResultHandler.java [hunk 1]
```diff
@@ -80,12 +84,31 @@ public RemotingCommand handle(final GetMessageResult getMessageResult,
...
+            subscriptionGroupConfig, response, clientAddress);
+        try {
+            processor.executeConsumeMessageHookBefore(request, requestHeader, getMessageResult, brokerAllowSuspend, response.getCode());
+        } catch (AbortProcessException e) {
+            response.setCode(e.getResponseCode());
+            response.setRemark(e.getErrorMessage());
+            return response;
+        }
 
+        //rewrite the response for the
```
Comment (lizhimins) [inline broker/src/main/java/org/apache/rocketmq/broker/processor/DefaultPullMessageResultHandler.java:100]: 注释不全

File: broker/src/main/java/org/apache/rocketmq/broker/processor/PopReviveService.java [hunk 2]
```diff
@@ -451,30 +465,79 @@ protected void mergeAndRevive(ConsumeReviveObj consumeReviveObj) throws Throwabl
...
+                .thenApply(messageExt -> {
+                    if (messageExt == null) {
+                        POP_LOGGER.warn("reviveQueueId={}, can not get biz msg topic is {}, offset is {}, then continue",
+                            queueId, popCheckPoint.getTopic(), msgOffset);
+                        return new Pair<>(msgOffset, true);
+                    }
+                    //skip ck from last epoch
+                    if (popCheckPoint.getPopTime() < messageExt.getStoreTimestamp()) {
+                        POP_LOGGER.warn("reviveQueueId={}, skip ck from last epoch {}", queueId, popCheckPoint);
+                        return new Pair<>(msgOffset, false);
```
Comment (xdkxlk) [inline broker/src/main/java/org/apache/rocketmq/broker/processor/PopReviveService.java:496]: In my opinion, there should be `new Pair<>(msgOffset, true)` to skip ck from last epoch

File: broker/src/main/java/org/apache/rocketmq/broker/failover/EscapeBridge.java [hunk 3]
```diff
@@ -283,6 +282,28 @@ public MessageExt getMessage(String topic, long offset, int queueId, String brok
         }
     }
 
+    public CompletableFuture<MessageExt> getMessageAsync(String topic, long offset, int queueId, String brokerName, boolean deCompressBody) {
```
Comment (lizhimins) [inline broker/src/main/java/org/apache/rocketmq/broker/failover/EscapeBridge.java:285]: 应该把同步接口也用异步的去实现

File: broker/src/main/java/org/apache/rocketmq/broker/out/BrokerOuterAPI.java [hunk 4]
```diff
@@ -1269,6 +1269,44 @@ public PullResult pullMessageFromSpecificBroker(String brokerName, String broker
         return pullResultExt;
     }
 
+    public CompletableFuture<PullResult> pullMessageFromSpecificBrokerAsync(String brokerName, String brokerAddr,
+                                                    String consumerGroup, String topic, int queueId, long offset,
+                                                    int maxNums,
+                                                    long timeoutMillis) throws RemotingException, InterruptedException {
```
Comment (lizhimins) [inline broker/src/main/java/org/apache/rocketmq/broker/out/BrokerOuterAPI.java:1275]: codestyle is strange

File: broker/src/main/java/org/apache/rocketmq/broker/failover/EscapeBridge.java [hunk 5]
```diff
@@ -283,6 +282,28 @@ public MessageExt getMessage(String topic, long offset, int queueId, String brok
...
+                    if (result == null) {
+                        LOG.warn("getMessageResult is null , innerConsumerGroupName {}, topic {}, offset {}, queueId {}", innerConsumerGroupName, topic, offset, queueId);
+                        return null;
+                    }
+                    List<MessageExt> list = decodeMsgList(result, deCompressBody);
+                    if (list == null || list.isEmpty()) {
+                        LOG.warn("Can not get msg , topic {}, offset {}, queueId {}, result is {}", topic, offset, queueId, result);
+                        return null;
+                    } else {
+                        return list.get(0);
```
Comment (aaron-ai) [inline broker/src/main/java/org/apache/rocketmq/broker/failover/EscapeBridge.java:299]: redundant else

File: broker/src/main/java/org/apache/rocketmq/broker/processor/PopReviveService.java [hunk 6]
```diff
@@ -122,7 +125,8 @@ private void reviveRetry(PopCheckPoint popCheckPoint, MessageExt messageExt) thr
         }
         if (putMessageResult.getAppendMessageResult() == null ||
             putMessageResult.getAppendMessageResult().getStatus() != AppendMessageStatus.PUT_OK) {
-            throw new Exception("reviveQueueId=" + queueId + ", revive error, msg is: " + msgInner);
+            POP_LOGGER.error("reviveQueueId=" + queueId + ", revive error, msg is: " + msgInner);
```
Comment (aaron-ai) [inline broker/src/main/java/org/apache/rocketmq/broker/processor/PopReviveService.java:128]: Use `{}` to print the argument.

File: store/src/main/java/org/apache/rocketmq/store/MessageStore.java [hunk 7]
```diff
@@ -137,6 +140,9 @@ GetMessageResult getMessage(final String group, final String topic, final int qu
     GetMessageResult getMessage(final String group, final String topic, final int queueId,
         final long offset, final int maxMsgNums, final int maxTotalMsgSize, final MessageFilter messageFilter);
 
+    CompletableFuture<GetMessageResult> getMessageAsync(final String group, final String topic, final int queueId,
```
Comment (aaron-ai) [inline store/src/main/java/org/apache/rocketmq/store/MessageStore.java:143]: Add comments for interface.

**Review by zhouxinyu — DISMISSED** [review]

**Review by aaron-ai — APPROVED** [review]

**Review by lizhimins — APPROVED** [review]

**Review by aaron-ai — APPROVED** [review]

**Pull request merged**